File: src/main/java/com/github/sorusclient/client/plugin/PluginManager.java
package com.github.sorusclient.client.plugin;

import org.apache.commons.io.IOUtils;
import org.json.JSONObject;

import java.io.File;
import java.io.IOException;
import java.net.URL;
import java.util.ArrayList;
import java.util.Enumeration;
import java.util.List;

public class PluginManager {

    private final List<Plugin> plugins = new ArrayList<>();

    public PluginManager() {
        try {
            Enumeration<URL> resources = PluginManager.class.getClassLoader().getResources("plugin.json");
            while (resources.hasMoreElements()) {
                URL resource = resources.nextElement();
                String contents = IOUtils.toString(resource.openStream());
                JSONObject json = new JSONObject(contents);

                String id = json.getString("id");
                String version = json.getString("version");
                String name = json.has("name") ? json.getString("name") : id;
                String description = json.has("description") ? json.getString("description") : "";

                String filePath = resource.toString();
                filePath = filePath.replace("jar:file:", "");
                filePath = filePath.substring(0, filePath.indexOf("!"));

                this.plugins.add(new Plugin(id, version, name, description, new File(filePath)));
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void remove(Plugin plugin) {
        plugin.getFile().delete();
        this.plugins.remove(plugin);
    }

    public List<Plugin> getPlugins() {
        return plugins;
    }

}

File: src/main/kotlin/com/github/sorusclient/client/plugin/PluginManager.kt
/*
 * Copyright (c) 2022. Danterus
 * Copyright (c) 2022. Sorus Contributors
 *
 * SPDX-License-Identifier: MPL-2.0
 */

package com.github.sorusclient.client.plugin

import com.github.sorusclient.client.adapter.AdapterManager
import com.github.sorusclient.client.bootstrap.BootstrapManager
import java.io.File

object PluginManager {

    private val plugins: MutableList<Plugin> = ArrayList()

    fun findPlugins() {
        val pluginFile = File("sorus/plugin")
        pluginFile.mkdirs()

        for (plugin in pluginFile.listFiles()!!) {
            BootstrapManager.addURL(plugin.toURI().toURL())
        }

        plugins.clear()

        for (resource in PluginManager::class.java.classLoader.getResources("plugin.json")) {
            val json = BootstrapManager.loadJson(resource, AdapterManager.getAdapter().version)

            val id = json.getString("id")

            if (plugins.any { it.id == id }) continue

            val version = json.getString("version")
            val name = if (json.has("name")) json.getString("name") else id
            val description = if (json.has("description")) json.getString("description") else ""
            val logo = if (json.has("logo")) json.getString("logo") else null
            var filePath = resource.toString()
            filePath = filePath.replace("jar:file:", "")
            filePath = filePath.substring(0, filePath.indexOf("!"))
            plugins.add(Plugin(id, version, name, description, File(filePath), logo))
        }
    }

    fun remove(plugin: Plugin) {
        plugin.file.delete()
        plugins.remove(plugin)
    }

    fun getPlugins(): List<Plugin> {
        return plugins
    }

}